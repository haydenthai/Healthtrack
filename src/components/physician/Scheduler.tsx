import React, { useEffect, useState } from 'react';
import {
    AppointmentPicker,
    AppointmentAttributesType,
} from 'react-appointment-picker';
import { calendar } from '../../util/schedulerMockData';

interface SchedulerProps {
    isDate: boolean;
    year: number;
    month: number;
    monday: number;
    i: number
}

function Scheduler({ isDate, year, month, monday, i }: SchedulerProps) {
    const [currentWeek, setCurrentWeek] = useState(calendar[i]);

    const physicianStartTime = 9;
    useEffect(() => {
        setCurrentWeek(calendar[i]);
    }, [currentWeek, i]);

    async function addAppointmentCallback({
        addedAppointment: { day, number, time, id },
        addCb,
    }: any): Promise<any> {

        // addCb changes the color of the box to green
        addCb(day, number, time, id);
        


        let newWeek = [...currentWeek];
        for (let i = 0; i < newWeek.length; i++) {
            for (let j = 0; j < newWeek[i].length; j++) {
                if (newWeek[i][j].id === id) {
                    newWeek[i][j].isSelected = true;

                    // newWeek[i][j].isReserved = true;

                }
            }
        }
        setCurrentWeek(newWeek);
    }

    async function removeAppointmentCallback(
        { day, number, time, id }: any,
        removeCb: (arg0: any, arg1: any) => void
    ): Promise<any> {

        //removeCb changes the color of the box from green to blue
        removeCb(day, number);
        let newWeek = [...currentWeek];

        for (let i = 0; i < newWeek.length; i++) {
            for (let j = 0; j < newWeek[i].length; j++) {
                if (
                    newWeek[i][j].id === id &&
                    newWeek[i][j].isSelected === true
                ) {
                    newWeek[i][j].isSelected = false;
                }
            }
        }
        setCurrentWeek(newWeek);
    }
    return (
        <div>
            <AppointmentPicker
                addAppointmentCallback={addAppointmentCallback}
                removeAppointmentCallback={removeAppointmentCallback}
                days={currentWeek as AppointmentAttributesType[][]}
                initialDay={new Date(year, month, monday, physicianStartTime)}
                unitTime={1_800_000}
                selectedByDefault={true}
                maxReservableAppointments={1000}
                local="en-US"
                alpha={isDate}
                visible={true}
            />
        </div>
    );
}

export default Scheduler;
