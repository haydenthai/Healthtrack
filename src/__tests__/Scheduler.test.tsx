import React from 'react';
import jsdom from 'jsdom';
import { createRoot } from 'react-dom/client';
import Scheduler from 'src/components/physician/Scheduler';

const { JSDOM } = jsdom;
const dom = new JSDOM('<!doctype html><html><body></body></html>');
global.document = dom.window.document;
global.window = dom.window as unknown as Window & typeof globalThis;

describe('<Scheduler />', () => {
    test('should render the element correctly', async () => {
        const container = document.createElement('div');
        createRoot(container).render(
            <Scheduler isDate={false} year={0} month={0} monday={0} i={0} />
        );
        expect(container).toMatchSnapshot();
    });

    test('should add appointment when clicking on a time slot', async () => {
        const container = document.createElement('div');
        createRoot(container).render(
            <Scheduler isDate={false} year={0} month={0} monday={0} i={0} />
        );
        const timeSlot = container.querySelector('.time-slot');
        timeSlot?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        expect(container).toMatchSnapshot();
    });

    test('should remove appointment when clicking on a time slot', async () => {
        const container = document.createElement('div');
        createRoot(container).render(
            <Scheduler isDate={false} year={0} month={0} monday={0} i={0} />
        );
        const timeSlot = container.querySelector('.time-slot');
        timeSlot?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        timeSlot?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        expect(container).toMatchSnapshot();
    });
});
