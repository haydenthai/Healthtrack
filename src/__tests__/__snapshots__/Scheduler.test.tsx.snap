// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scheduler /> should add appointment when clicking on a time slot 1`] = `<div />`;

exports[`<Scheduler /> should remove appointment when clicking on a time slot 1`] = `<div />`;

exports[`<Scheduler /> should render the element correctly 1`] = `<div />`;
